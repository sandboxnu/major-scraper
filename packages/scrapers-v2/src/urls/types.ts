export enum College {
  ARTS_MEDIA_DESIGN = "arts-media-design",
  BUSINESS = "business",
  KHOURY = "computer-information-science",
  ENGINEERING = "engineering",
  HEALTH_SCIENCES = "health-sciences",
  SCIENCE = "science",
  SOCIAL_SCIENCES_HUMANITIES = "social-sciences-humanities",
}

type CollegeKeys = keyof typeof College;

export type AvailableMajors = {
  [key in CollegeKeys]: Array<string>;
};
