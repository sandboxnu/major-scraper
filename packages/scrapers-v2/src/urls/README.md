# URL scraper

This folder contains code for scraping the URLs of entries in the catalog, and is based on the sidebar _hierarchy_ that the catalog uses to describe the locations of catalog entries. 

You can think of the catalog sidebar as a bunch of entries, with each entry potentially containing more sub-entries (hence _hierarchical_ in nature).

## Scraping Strategy

We hardcode a set of "top level entries" to scrape, which are the colleges. They are listed underneath [undergraduate](https://catalog.northeastern.edu/undergraduate/), and are just the colleges: CAMD, D'Amore McKim, Khoury, COE, Bouv√©, COE, and CSSH.
 
When you click on one of the colleges above, (from the undergraduate link) you can see the sidebar will expand, with more entries. We then add all of these sub-entries to a queue of links to visit, and begin by visiting the first (think, BFS).

Visiting the first one, "School of Architecture" (under CAMD), we see it yields more links. Rinse and repeat the above process. Visiting the first of these, we see that this time, it does _not_ yield more sub-links. Therefore, we have a "leaf" of the hierarchy "tree", and we store the link to this entry.

## Misc

When actually scraping, we just store a list of paths (each path to a catalog entry). However because of the hierarchical nature of the catalog, we figured it may be useful to later be able to have a hierarchy, so we convert it to be a hierarchy.

Maybe in the future this is used, or maybe it is not needed and can be removed.

## Support for old catalogs

All the old catalogs can be found here: https://registrar.northeastern.edu/group/academic-catalogs/

However, only the last ~7 or so catalogs are actually stored in HTML form. we don't know if earlier versions were in HTML at one point, but the server was taken offline, or if there was only ever a PDF of the catalog available.

Because of this, in the future, it may be worth it to download old versions of the catalog for safekeeping.

Also because of this, we cannot scrape older versions of the catalog (older than 7 years), because the scraper cannot function on non-HTML versions of the catalog. 

The ticket for adding old catalog support can be found here: https://trello.com/c/7aFTWtMU/452-scrapers-add-old-catalog-support