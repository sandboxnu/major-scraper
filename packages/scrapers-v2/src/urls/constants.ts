export const ALL_COLLEGES = [
  "arts-media-design",
  "business",
  "computer-information-science",
  "engineering",
  "health-sciences",
  "science",
  "social-sciences-humanities",
];

/*
Master list of old catalogs:
https://registrar.northeastern.edu/group/academic-catalogs/

Catalogs that are scrape-able (HTML):
2021–2022 Catalog and Course Descriptions (current year)
2020–2021 Catalog and Course Descriptions
2019–2020 Catalog and Course Descriptions
2018–2019 Catalog and Course Descriptions
2017–2018 Catalog and Course Descriptions
2016–2017 Catalog and Course Descriptions
*/
