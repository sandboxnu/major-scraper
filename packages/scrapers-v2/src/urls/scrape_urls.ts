import { appendPath, loadHTML } from "../utils";
import { CatalogHierarchy, College, CatalogPath } from "./types";

export const scrapeMajorLinks = async (start: number, end: number) => {
  if (start !== end - 1) {
    throw new Error("start should == end-1");
  }

  if (start < 2016) {
    // this is because there is no HTML version of those catalogs
    throw new Error("scraping for years before 2016-2017 are not supported");
  }

  if (start !== 2021) {
    // todo: implement generic scraping from overall archive
    // https://registrar.northeastern.edu/group/academic-catalogs/
    throw new Error("only current year is supported");
  }

  return scrapeMajorLinksForUrl(
    "https://catalog.northeastern.edu",
    "undergraduate"
  );
};

/**
 * Given a baseUrl and a path, attempts to scrape the major catalog based on the sidebar
 * hierarchy.
 * @param baseUrl the base url of the major catalog. should look something
 *                like "https://catalog.northeastern.edu"
 * @param path    the path of the major catalog. something like "/undergraduate"
 *                or "archive/2018-2019/undergraduate/". trailing or leading slashes are ok.
 */
export const scrapeMajorLinksForUrl = async (
  baseUrl: string,
  path: string
): Promise<CatalogHierarchy> => {
  const paths = getPathParts(path);
  try {
    const initStack = Object.values(College).map((college) => ({
      path: [...paths, college],
    }));
    const catalogPaths = await scrapeLinks(baseUrl, initStack);
    return convertToHierarchy(baseUrl, catalogPaths);
  } catch (e) {
    throw e;
  }
};

const scrapeLinks = async (
  baseUrl: string,
  initQueue: CatalogPath[]
): Promise<CatalogPath[]> => {
  const done: CatalogPath[] = [];

  // avoid minors?
  let queue = initQueue;
  while (queue.length > 0) {
    const pages = await Promise.all(getUrlHtmls(queue, baseUrl));
    const nextQueue: CatalogPath[] = [];
    for (const { $, url } of pages) {
      const children = getChildrenForPathId($, url).toArray().map($);
      for (const element of children) {
        const path = getLinkForEl(element);
        const bucket = isParent(element) ? nextQueue : done;
        bucket.push({ path });
      }
    }
    queue = nextQueue;
  }

  return done;
};

const convertToHierarchy = (
  base: string,
  catalogPaths: CatalogPath[]
): CatalogHierarchy => {
  const hierarchy: CatalogHierarchy = {};
  const addPathToHierarchy = (path: string[]) => {
    let obj: CatalogHierarchy = hierarchy;
    for (let i = 0; i < path.length - 1; i += 1) {
      const part = path[i];
      if (!(part in obj)) {
        obj[part] = {};
      }
      const child = obj[part];
      if (typeof child === "string") {
        throw new Error(
          "Hierarchy was inconsistent: found a child, where a parent was expected"
        );
      }
      obj = child;
    }
    const last = path[path.length - 1];
    obj[last] = joinParts(base, path).toString();
  };
  for (const { path } of catalogPaths) {
    addPathToHierarchy(path);
  }
  return hierarchy;
};

const isParent = (el: Cheerio) => {
  return el.hasClass("isparent");
};

const getPathParts = (url: string) => {
  return url.split("/").filter((s) => s !== "");
};

const getLinkForEl = (element: Cheerio): string[] => {
  const aTag = element.find("a");
  if (aTag.length === 0) {
    const msg = "Catalog is missing a link for a parent node";
    throw new Error(msg);
  }

  return getPathParts(aTag.attr("href"));
};

const getChildrenForPathId = ($: CheerioStatic, url: URL) => {
  const id = url.pathname.split("/").join("\\/");
  const current = $(`#${id}\\/`);
  return current.children();
};

const fetchUrlHtml = (url: URL) =>
  new Promise<{ $: CheerioStatic; url: URL }>((res) => {
    loadHTML(url.href).then((r) => res({ $: r, url }));
  });

const joinParts = (base: string, parts: string[]) => {
  return appendPath(base, parts.join("/"));
};

const getUrlHtmls = (queue: CatalogPath[], base: string) => {
  return queue.map(({ path }) => joinParts(base, path)).map(fetchUrlHtml);
};
