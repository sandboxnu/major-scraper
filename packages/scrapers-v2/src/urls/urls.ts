import { convertToHierarchy, joinParts, loadHTML } from "../utils";
import { CatalogHierarchy, College, CatalogPath } from "./types";

/**
 * Scrapes all catalog entries underneath the colleges for the specified catalog
 * year (given in the form of two numbers to avoid ambiguity: ex, 2021-2022).
 *
 * @param start starting year (must be end year - 1)
 * @param end   ending year
 * @returns     a hierarchy of catalog entry links
 */
export const scrapeMajorLinks = async (
  start: number,
  end: number
): Promise<CatalogHierarchy> => {
  if (start !== end - 1) {
    throw new Error("start should == end-1");
  }

  if (start < 2016) {
    // this is because there is no HTML version of those catalogs
    throw new Error("scraping for years before 2016-2017 are not supported");
  }

  if (start !== 2021) {
    // todo: implement generic scraping from overall archive
    // https://registrar.northeastern.edu/group/academic-catalogs/
    throw new Error("only current year is supported");
  }

  return scrapeMajorLinksForUrl(
    "https://catalog.northeastern.edu",
    "undergraduate"
  );
};

/**
 * Given a baseUrl and a path, attempts to scrape the major catalog based on the sidebar
 * hierarchy.
 * @param baseUrl the base url of the major catalog. should look something
 *                like "https://catalog.northeastern.edu"
 * @param path    the path of the major catalog. something like "/undergraduate"
 *                or "archive/2018-2019/undergraduate/". trailing or leading slashes are ok.
 */
export const scrapeMajorLinksForUrl = async (
  baseUrl: string,
  path: string
): Promise<CatalogHierarchy> => {
  const paths = getPathParts(path);
  const initQueue = Object.values(College).map((college) => ({
    path: [...paths, college],
  }));
  const catalogPaths = await scrapeLinks(baseUrl, initQueue);
  return convertToHierarchy(baseUrl, catalogPaths);
};

/**
 *
 * Retrieves all sub-entries of the given initial queue in BFS fashion using the catalog sidebar hierarchy.
 *
 * @param baseUrl   the base catalog URL, i.e. https://catalog.northeastern.edu
 * @param initQueue a queue of parent entries
 * @returns         a flat list of all the last level children catalog entries
 */
const scrapeLinks = async (
  baseUrl: string,
  initQueue: CatalogPath[]
): Promise<CatalogPath[]> => {
  const results: CatalogPath[] = [];

  let queue = initQueue;
  while (queue.length > 0) {
    const pages = await Promise.all(getUrlHtmls(queue, baseUrl));
    const nextQueue: CatalogPath[] = [];
    for (const { $, url } of pages) {
      const children = getChildrenForPathId($, url).toArray().map($);
      for (const element of children) {
        const path = getLinkForEl(element);
        const bucket = isParent(element) ? nextQueue : results;
        bucket.push({ path });
      }
    }
    queue = nextQueue;
  }

  return results;
};

const isParent = (el: Cheerio) => {
  return el.hasClass("isparent");
};

const getPathParts = (url: string) => {
  return url.split("/").filter((s) => s !== "");
};

const getLinkForEl = (element: Cheerio): string[] => {
  const aTag = element.find("a");
  if (aTag.length === 0) {
    const msg = "Catalog is missing a link for a parent element.";
    throw new Error(msg);
  }

  return getPathParts(aTag.attr("href"));
};

const getChildrenForPathId = ($: CheerioStatic, url: URL) => {
  // The catalog entries have an ID equal to the path, with a trailing slash
  // We select the element via its ID
  // Note: for getElementById, forward slashes need to be escaped
  const id = url.pathname.split("/").join("\\/");
  const current = $(`#${id}\\/`);
  return current.children();
};

const fetchUrlHtml = async (url: URL) => {
  const r = await loadHTML(url.href);
  return { $: r, url };
};

const getUrlHtmls = (queue: CatalogPath[], base: string) => {
  return queue.map(({ path }) => joinParts(base, path)).map(fetchUrlHtml);
};
