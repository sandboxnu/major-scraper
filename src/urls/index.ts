export * from "./urls";
export * from "./types";
