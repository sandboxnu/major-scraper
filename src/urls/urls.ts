import { loadHTML2, loadHtmlWithUrl } from "../utils";
import { CatalogURLResult, College } from "./types";
import { ResultType } from "../graduate-types/common";
import { join } from "path";
import {
  BASE_URL,
  CURRENT_CATALOG_YEAR,
  EARLIEST_CATALOG_YEAR,
} from "../constants";

/**
 * Scrapes all catalog entries underneath the colleges for the specified catalog
 * year (given in the form of two numbers to avoid ambiguity: ex, 2021-2022).
 *
 * @param   start Starting year (must be end year - 1)
 * @param   end   Ending year
 * @returns       A hierarchy of catalog entry links
 */
export const scrapeMajorLinks = async (
  start: number,
): Promise<CatalogURLResult> => {
  if (start < EARLIEST_CATALOG_YEAR) {
    throw new Error("Scraping for years before 2016-2017 is not supported.");
  } else if (start > CURRENT_CATALOG_YEAR) {
    throw new Error(
      "Either you're scraping for a year in the future (which won't work unless time travel has been invented since this message was written), or you need to update CURRENT_CATALOG_YEAR in constants.ts.",
    );
  }

  const path =
    start == CURRENT_CATALOG_YEAR
      ? "undergraduate"
      : `archive/${start}-${start + 1}/undergraduate`;

  const initQueue = Object.values(College).map(
    college => new URL(join(BASE_URL, path, college, "/")),
  );

  return await scrapeLinks(BASE_URL, initQueue);
};

/**
 * Retrieves all sub-entries of the given initial queue in BFS fashion using the
 * catalog sidebar hierarchy.
 *
 * @param   baseUrl   The base catalog URL, i.e. https://catalog.northeastern.edu
 * @param   initQueue A queue of parent entries
 * @returns           A flat list of all the last level children catalog entries
 */
const scrapeLinks = async (
  baseUrl: string,
  initQueue: URL[],
): Promise<CatalogURLResult> => {
  const entries: URL[] = [];
  const unfinished = [];

  // there are multiple links in the sidebar to the same entry
  // keep a set to avoid visiting the same entry twice
  const seen = new Set(initQueue.map(url => url.href));
  let queue = initQueue;
  while (queue.length > 0) {
    const { ok, errors } = await getUrlHtmls(queue);
    unfinished.push(...errors);

    const nextQueue: URL[] = [];
    for (const { $, url } of ok) {
      const children = getChildrenForPathId($, url).toArray().map($);

      for (const element of children) {
        const path = getLinkForEl(element);
        const url = new URL(join(baseUrl, path));

        if (
          !seen.has(url.href) &&
          Object.values(College).some(linkPart => url.href.includes(linkPart))
        ) {
          const bucket = isParent(element) ? nextQueue : entries;
          bucket.push(url);
          seen.add(url.href);
        }
      }
    }
    queue = nextQueue;
  }

  return { entries, unfinished };
};

const isParent = (el: Cheerio) => {
  return el.hasClass("isparent");
};

const getLinkForEl = (element: Cheerio) => {
  const aTag = element.find("a");
  if (aTag.length === 0) {
    const msg = "Catalog is missing a link for a parent element.";
    throw new Error(msg);
  }

  return aTag.attr("href");
};

const getChildrenForPathId = ($: CheerioStatic, url: URL) => {
  // The catalog entries have an ID equal to the path, with a trailing slash
  // We select the element via its ID
  // Note: for getElementById, forward slashes need to be escaped
  let id = url.pathname;

  // Remove archive url prefix if present
  id = id.replaceAll(/^\/archive\/20\d\d-20\d\d/g, "");
  id = id.replaceAll("/", "\\/");

  const current = $(`#${id}`);
  return current.children();
};

const getUrlHtmls = async (queue: URL[]) => {
  const fetchResults = await Promise.all(queue.map(loadHTML2));

  const ok = [];
  const errors = [];
  for (const { url, result } of fetchResults) {
    if (result.type === ResultType.Ok) {
      ok.push({ $: result.ok, url });
      continue;
    }
    errors.push({ error: result.err, url });
  }
  return { ok, errors };
};
