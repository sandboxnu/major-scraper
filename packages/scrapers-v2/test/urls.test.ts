// https://catalog.northeastern.edu/undergraduate

import { scrapeMajorLinks } from "../src/urls/scrape_urls";

describe("url scraper", () => {
  test("scrape", async () => {

    const result = await scrapeMajorLinks(2021, 2022);
    expect(result).toMatchSnapshot();
  });
});
