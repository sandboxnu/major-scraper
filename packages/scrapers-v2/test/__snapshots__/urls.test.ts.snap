// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`url scraper scrape 1`] = `
Object {
  "ARTS_MEDIA_DESIGN": Array [],
  "BUSINESS": Array [],
  "ENGINEERING": Array [],
  "HEALTH_SCIENCES": Array [],
  "KHOURY": Array [],
  "SCIENCE": Array [],
  "SOCIAL_SCIENCES_HUMANITIES": Array [],
}
`;
